docker/README.md

```diff
@@ -0,0 +1,42 @@
+## Docker images
+
+### Installing
+
+The Docker image may be used for installing spyro. The following command builds the release image:
+````
+docker build -t runtag:1.0 --target spyro_release .
+````
+
+Then, the following commands gives access to a virtual environment with spyro:
+````
+docker run -it runtag:1.0
+. firedrake/bin/activate
+````
+
+### Testing
+
+The Docker image for running the tests is built with the following command (considering that your working directory is the same where the Dockerfile is located):
+````
+docker build -t testtag:1.0 --target spyro_test .
+````
+
+Then, the following command may be called for running the tests:
+````
+docker run -it testtag:1.0 /bin/bash -c "source /home/firedrake/firedrake/bin/activate; python3 -m pytest --maxfail=1 ."
```

Comment:
The tests are currently failing inside the docker. I am going to try to figure out why

Reply:
The first test to fail produces the following log (tested in firedrake tags 2024-05, 2024-04, and 2024-03):

```
shared/spyro/test/test_MMS.py:36: in run_solve
    Wave_obj.forward_solve()
shared/spyro/spyro/solvers/acoustic_wave.py:46: in forward_solve
    self.wave_propagator()
shared/spyro/spyro/io/basicio.py:86: in wrapper
    u, u_r = func(*args, **dict(kwargs, source_num=snum))
shared/spyro/spyro/solvers/acoustic_wave.py:113: in wave_propagator
    usol, usol_recv = time_integrator(self, source_id=source_num)
shared/spyro/spyro/solvers/time_integration.py:10: in time_integrator
    return time_integrator_mms(Wave_object, source_id=source_id)
shared/spyro/spyro/solvers/time_integration.py:24: in time_integrator_mms
    return central_difference_MMS(Wave_object, source_id=source_id)
shared/spyro/spyro/solvers/time_integration_central_difference.py:280: in central_difference_MMS
    Wave_object.solver.solve(X, B)
petsc4py/PETSc/Log.pyx:188: in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    ???
petsc4py/PETSc/Log.pyx:189: in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    ???
firedrake/src/firedrake/firedrake/linear_solver.py:159: in solve
    b = self._lifted(b)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <firedrake.linear_solver.LinearSolver object at 0x7f787f451d20>
b = Coefficient(WithGeometry(FunctionSpace(<firedrake.mesh.MeshTopology object at 0x7f7889c7aef0>, FiniteElement('Q', quadrilateral, 4, variant='spectral'), name=None), Mesh(VectorElement(FiniteElement('Q', quadrilateral, 1), dim=2), 1)), 31)

    def _lifted(self, b):
        u, update, blift = self._rhs
        u.dat.zero()
        for bc in self.A.bcs:
            bc.apply(u)
        update()
        # blift contains -A u_bc
>       blift += b
E       TypeError: unsupported operand type(s) for +=: 'Cofunction' and 'Function'

firedrake/src/firedrake/firedrake/linear_solver.py:129: TypeError
```

Reply:
Since this error has been present, it has been present for the last three months of Firedrake. It's either a long-standing bug or a deprecated feature. Could you check 2023-12? My latest local Firedrake install was 2023-05, but yours should be newer. For now, we need a stable version to use with our tests.

I am currently installing the latest Firedrake locally to test Spyro and check if it reproduces the same error as the docker. I have also done a quick look at the open bugs in the Firedrake repo to see if any match ours. After that, I plan on figuring out exactly what causes the error and submitting a minimal code with it as a bug in the repo.

Reply:
I'm running a test with 2023-08 (the oldest image in DockerHub with the exception of experimental)  and this error has not happened yet (the `test_MMS.py` is still running). As this error happens before the code actually runs, I believe that version 2023-08 will not have this issue.

Reply:
I am running the tests locally using the latest version. Not only is there an ¨unsupported operand type(s) for +=: 'Cofunction' and 'Function'¨ error in the MMS test, but there are numerical instabilities and more significant errors in other forward tests that go through. My initial guess is that something broke the mass lumped triangular elements in Firedrake and since we are probably the only ones that use them no one noticed it until now. They still build diagonal mass matrices and pass Firedrake's unit tests. I still have more tests to run to diagnose if this is the issue. If it is, I will try to submit an issue and/or a fix for it in their GitHub (since I already have Firedrake code for some new ML elements that I need to submit and am familiar with the relevant FIAT and TSFC files).

Reply:
Now the test `test_cpw_calc.py`is failing because the file `reference_solution_cpw.npy` is missing. We have to define where to store these files and how to access during development and testing.

```
FAILED shared/spyro/test/test_cpw_calc.py::test_cpw_calc - FileNotFoundError: [Errno 2] No such file or directory: 'test/inputfiles/reference_solution_cpw.npy'
```

Reply:
I am running tests to check what is happening. This test has an input file located in test/inputfiles/ and loads it when I run it locally. It does not use any file not found in the repository

Reply:
I think I have discovered the causes of the errors in the new Firedrake installation. We can actually bypass the solver error by using a cofunction in the dual of the function space on the right-hand side. However, there will still be a new issue correlated with sum factorized elements (such as the Gauss Lobato Legendre quads and hexas). The spectral elements take longer in the new Firedrake and produce greater error. The mass-lumped triangular elements also have greater numerical instability. 

So until this is solved we should not update Firedrake

Reply:
Unfortunately, you do not have access to an older Firedrake version. I had a look at all the available Firedrakes after 2023-08, and they all are very slow and have greater errors overall.

Reply:
I think the problem is also related to our LinearSolver object. I am going to try, in the compatibility branch, to update the RHS everywhere to cofunctions and to change our solver object to LinearVariationalSolver. It is possible that this will make the code compatible with newer versions of Firedrake

Reply:
> Unfortunately, you do not have access to an older Firedrake version. I had a look at all the available Firedrakes after 2023-08, and they all are very slow and have greater errors overall.

I think we can install older versions (such as 'https://github.com/firedrakeproject/firedrake/tree/Firedrake_20221208.0'). It may be more complex to install them with Docker, but at least we will we have the same performance as before.

Reply:
I noticed that I skipped testing for 2023-09 before. I retested using all available Firedrake Docker versions, and found that the 2023-09 version passes all tests and performs better than the 2023-08 version. I suggest we prioritize adjusting the Docker file to focus on this version for now.

Reply:
If version 2023-09 is working, it would save us a lot of time. I was trying to install version `Firedrake_20230316.0`, but the process is slow because the install script does not work out of the box with Docker and some errors happen after the PETSc installation (which is slow). Can I use the Docker image of version 2023-09 as our base installation?

Reply:
That sounds like a good plan. Using the Docker image of version 2023-09 should streamline the process and save us time. Installing older versions and managing dependencies can be really annoying. Let's go with 2023-09 for now. We can revisit and update to the latest Firedrake version when we are compatible.